src/components/token/TokenBalances.js

```diff
@@ -0,0 +1,327 @@
+    });
+
+    return get(tokenBalanceInformationRequest, 'data', {
+      transactions: 0,
+      addresses: 0,
+    });
+  }
+
+  /**
+   * Calls elastic search with state data, sets loading and URL information
+   */
+  performSearch = async () => {
+    this.setState({ isSearchLoading: true });
+    const tokenBalances = await this.getTokenBalances([]);
+    const tokenBalanceInformation = await this.loadTokenBalanceInformation();
+
+    this.setState({
+      isSearchLoading: false,
+      page: 1,
+      tokenBalances: tokenBalances.hits,
+      transactionsCount: tokenBalanceInformation.transactions,
+      addressesCount: tokenBalanceInformation.addresses,
+      hasAfter: tokenBalances.has_next,
+      hasBefore: false,
+      pageSearchAfter: [{
+        page: 1,
+        searchAfter: []
+      }],
+    });
+
+    // This is ultimately called when search text, sort, or sort order changes
+    this.updateURL();
+  }
+
+  /**
+   * Update the URL, so user can share the results of a search
+   */
+  updateURL = () => {
+    const newURL = this.pagination.setURLParameters({
+      sortBy: this.state.sortBy,
+      order: this.state.order,
+    });
+
+    this.props.history.push(newURL);
+  }
+
+  /**
+    * Process events when next page is requested by user
+    */
+  nextPageClicked = async () => {
+    this.setState({ calculatingPage: true });
+
+    const nextPage = this.state.page + 1;
+    let searchAfter = get(find(this.state.pageSearchAfter, { page: nextPage }), 'searchAfter', []);
+
+    // Calculate searchAfter of next page if not already calculated
+    if (isEmpty(searchAfter)) {
+      const lastCurrentTokenSort = get(last(this.state.tokenBalances), 'sort', []);
+
+      const newEntry = {
+        page: nextPage,
+        searchAfter: lastCurrentTokenSort,
+      }
+
+      this.setState({
+        pageSearchAfter: [
+          ...this.state.pageSearchAfter,
+          newEntry,
+        ],
+      });
+
+      searchAfter = lastCurrentTokenSort;
+    }
+
+    const tokenBalances = await this.getTokenBalances(searchAfter);
+
+    this.setState({
+      tokenBalances: tokenBalances.hits,
+      hasAfter: tokenBalances.has_next,
+      hasBefore: true,
+      page: nextPage,
+      calculatingPage: false,
+    });
+  }
+
+  /**
+   * Process events when previous page is requested by user
+   */
+  previousPageClicked = async () => {
+    this.setState({ calculatingPage: true });
+
+    const previousPage = this.state.page - 1;
+    const searchAfter = get(find(this.state.pageSearchAfter, { page: previousPage }), 'searchAfter', []);
+    const tokenBalances = await this.getTokenBalances(searchAfter);
+
+    this.setState({
+      tokenBalances: tokenBalances.hits,
+      hasAfter: true,
+      hasBefore: previousPage === 1 ? false : true,
+      page: previousPage,
+      calculatingPage: false,
+    });
+  }
+
+  onTokenSelected = async (token) => {
+    if (!token) {
+      await this.setState({
```

Comment:
Does it make sense to await `this.setState`? I think it doesn't return a Promise to be awaited

Reply:
You are right, refactored in c67608b9dd1fe17265aecc17822d0fe19e1af2bf